File: Module - Services/Services2/app/src/main/java/com/javadude/services2/RemoteService1.java
package com.javadude.services2;

import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import androidx.annotation.Nullable;
import android.util.Log;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class RemoteService1 extends Service {
	public static final int RESET = 1;
	public static final int ADD_REPORTER = 2;
	public static final int REMOVE_REPORTER = 3;
	public static final int REPORT = 100;
	public void reset() {
		i = 1;
	}

	private static class MyHandler extends Handler {
		private WeakReference<RemoteService1> serviceRef;

		public MyHandler(RemoteService1 remoteService1) {
			serviceRef = new WeakReference<>(remoteService1);
		}

		@Override
		public void handleMessage(Message msg) {
			RemoteService1 remoteService1 = serviceRef.get();
			if (remoteService1 != null) {
				switch(msg.what) {
					case RESET:
						remoteService1.reset();
						break;
					case ADD_REPORTER:
						remoteService1.reporters.add(msg.replyTo);
						break;
					case REMOVE_REPORTER:
						remoteService1.reporters.remove(msg.replyTo);
						break;
					default:
						super.handleMessage(msg);
				}
			}
		}
	}
	private List<Messenger> reporters = new ArrayList<>();
	private Messenger messenger = new Messenger(new MyHandler(this));


	@Override
	public void onCreate() {
		super.onCreate();
		Log.d("StartedService", "onCreate");
	}

	private volatile int i = 1;
	private class CounterThread extends Thread {
		@Override public void run() {
			for(i = 1; !isInterrupted() && i <=100; i++) {
				Log.d("RemoteService1", "count = " + i);
				Message message = Message.obtain(null, REPORT, i, 0);
				for(Messenger messenger : reporters) {
					try {
						messenger.send(message);
					} catch (RemoteException e) {
						Log.e(getClass().getSimpleName(), "Error sending report", e);
					}
				}
				try {
					Thread.sleep(500);
				} catch (InterruptedException e) {
					interrupt();
				}
			}
			stopSelf();
		}
	}

	private CounterThread counterThread;

	@Override
	public void onDestroy() {
		super.onDestroy();
		counterThread.interrupt();
		Log.d("RemoteService1", "onDestroy");
	}

	@Nullable @Override public synchronized IBinder onBind(Intent intent) {
		if (counterThread == null) {
			counterThread = new CounterThread();
			counterThread.start();
		}
		return messenger.getBinder();
	}

	@Override
	public boolean onUnbind(Intent intent) {
		Log.d("RemoteService1", "onUnbind");
		return super.onUnbind(intent);
	}
}

File: Module - Services/Services2/app/src/main/java/com/javadude/services2/RemoteService1.kt
package com.javadude.services2

import android.app.Service
import android.content.Intent
import android.os.*
import android.util.Log
import java.lang.ref.WeakReference

// implementation of our Messenger/Handler-based remote service
class RemoteService1 : Service() {
    private val reporters = mutableListOf<Messenger>()
    private val messenger = Messenger(MyHandler(this))
    @Volatile var i = 1

    fun reset() {
        i = 1
    }
    fun addReporter(reporter : Messenger) {
        reporters.add(reporter)
    }
    fun removeReporter(reporter : Messenger) {
        reporters.remove(reporter)
    }

    private class MyHandler(remoteService1: RemoteService1) : Handler(Looper.getMainLooper()) {
        private val serviceRef = WeakReference(remoteService1)

        override fun handleMessage(msg: Message) {
            serviceRef.get()?.let {
                when (msg.what) {
                    RESET -> it.reset()
                    ADD_REPORTER -> it.addReporter(msg.replyTo)
                    REMOVE_REPORTER -> it.removeReporter(msg.replyTo)
                    else -> super.handleMessage(msg)
                }
            }
        }
    }

    private var counterThread : Thread? = null

    private fun createCounter() = object : Thread() {
        override fun run() {
            while(!isInterrupted && i <= 100) {
                Log.d("StartedService", "count = $i")
                val message = Message.obtain(null, REPORT, i, 0)
                reporters.forEach {
                    it.send(message)
                }
                try {
                    sleep(250)
                } catch (e : InterruptedException) {
                    interrupt()
                }

                i++
            }
            stopSelf()
        }
    }

    override fun onDestroy() {
        counterThread?.interrupt()
        super.onDestroy()
    }

    override fun onBind(p0: Intent?): IBinder? {
        if (counterThread == null) {
            counterThread = createCounter()
            counterThread!!.start()
        }
        return messenger.binder
    }

    companion object {
        const val RESET = 1
        const val ADD_REPORTER = 2
        const val REMOVE_REPORTER = 3
        const val REPORT = 100
    }
}
